main/modes/music/usbsynth/mode_synth.c

```diff
@@ -0,0 +1,956 @@
+    c505, // A# - Fuchsia
+    c503, // B  - Pink
+};
+
+const char synthModeName[] = "USB MIDI Synth";
+
+swadgeMode_t synthMode = {
+    .modeName                 = synthModeName,
+    .wifiMode                 = NO_WIFI,
+    .overrideUsb              = true,
+    .usesAccelerometer        = false,
+    .usesThermometer          = false,
+    .overrideSelectBtn        = false,
+    .fnEnterMode              = synthEnterMode,
+    .fnExitMode               = synthExitMode,
+    .fnMainLoop               = synthMainLoop,
+    .fnAudioCallback          = NULL,
+    .fnBackgroundDrawCallback = NULL,
+    .fnEspNowRecvCb           = NULL,
+    .fnEspNowSendCb           = NULL,
+    .fnAdvancedUSB            = NULL,
+    .fnDacCb                  = synthDacCallback,
+};
+
+static synthData_t* sd;
+
+// Functions
+
+static void synthEnterMode(void)
+{
+    sd = calloc(1, sizeof(synthData_t));
+    loadFont("ibm_vga8.font", &sd->font, true);
+    sd->perc[9] = true;
+    midiPlayerInit(&sd->midiPlayer);
+    sd->noteTime           = 200000;
+    sd->pitch              = 0x2000;
+    sd->longestProgramName = gmProgramNames[24];
+    sd->nextExpiry         = 200000;
+
+    sd->fileMode = true;
+    if (sd->fileMode)
+    {
+        if (loadMidiFile("stereo.mid", &sd->midiFile, true))
+        {
+            sd->midiPlayer.textMessageCallback = midiTextCallback;
+            midiSetFile(&sd->midiPlayer, &sd->midiFile);
+            midiPause(&sd->midiPlayer, false);
+            sd->midiPlayer.loop = true;
+        }
+        else
+        {
+            ESP_LOGE("Synth", "Could not load MIDI file");
+        }
+    }
+    else
+    {
+        sd->installed                    = installMidiUsb();
+        sd->midiPlayer.streamingCallback = usbMidiCallback;
+        sd->midiPlayer.mode              = MIDI_STREAMING;
+        midiPause(&sd->midiPlayer, false);
+    }
+
+    loadWsg("piano.wsg", &sd->instrumentImages[0], true);
+    loadWsg("chromatic_percussion.wsg", &sd->instrumentImages[1], true);
+    loadWsg("organ.wsg", &sd->instrumentImages[2], true);
+    loadWsg("guitar.wsg", &sd->instrumentImages[3], true);
+    loadWsg("bass.wsg", &sd->instrumentImages[4], true);
+    loadWsg("solo_strings.wsg", &sd->instrumentImages[5], true);
+    loadWsg("ensemble.wsg", &sd->instrumentImages[6], true);
+    loadWsg("brass.wsg", &sd->instrumentImages[7], true);
+    loadWsg("reed.wsg", &sd->instrumentImages[8], true);
+    loadWsg("pipe.wsg", &sd->instrumentImages[9], true);
+    loadWsg("synth_lead.wsg", &sd->instrumentImages[10], true);
+    loadWsg("synth_pad.wsg", &sd->instrumentImages[11], true);
+    loadWsg("synth_effects.wsg", &sd->instrumentImages[12], true);
+    loadWsg("ethnic.wsg", &sd->instrumentImages[13], true);
+    loadWsg("percussive.wsg", &sd->instrumentImages[14], true);
+    loadWsg("sound_effects.wsg", &sd->instrumentImages[15], true);
+    loadWsg("percussion.wsg", &sd->percussionImage, true);
+
+    // MAXIMUM SPEEEEED
```

Comment:
Truly the year of "gotta go fast." Also why does this need a high framerate?

Reply:
Ah, it actually doesn't! This is from when the USB read was in the main loop but that's now part of the MIDI player loop so now this is just for faster graphics updates.